proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,671 @@
+doubt invent many more use cases over time.
+
+Proposed Change Specification
+-----------------------------
+
+We introduce a new language extension. Types with a linearity
+specification are syntactically legal anywhere in a module if and only
+if ``-XLinearTypes`` is turned on.
+
+Definition
+~~~~~~~~~~
+
+We say that a function ``f`` is *linear* when ``f u`` is consumed
+exactly once implies that ``u`` is *consumed exactly once* (defined
+as follows).
+
+- Consuming a value of a data type exactly once means evaluating it to
+  head normal form, then consuming its fields exactly once
+- Consuming a function exactly once means applying it and consuming
+  its result exactly once
+
+The type of linear function from type ``A`` to type ``B`` is written
+``A ->. B`` (see syntax below).
+
+Linearity is a strengthening of the contract of the regular function
+type ``A -> B``, which will be called the type of *unrestricted*
+functions.
+
+Remark: linear function ``f`` can diverge or be called on diverging
+data. It may feel weird because ``f`` will not necessarily consume its
+argument. But it's alright: the metatheory still works. One can think
+of a diverging computation as not consuming its result exactly once,
+therefore, it is vacuously true that if ``f u`` is consumed exactly
+once, then ``u`` is consumed exactly once.
+
+Polymorphism
+~~~~~~~~~~~~
+
+In order for linear functions and unrestricted functions not to live
+in completely distinct worlds, to avoid code duplication, we
+introduce a notion of polymorphism, dubbed *multiplicity polymorphism*,
+over whether a function is linear.
+
+A linear function is said to have multiplicity ``1`` while an
+unrestricted function is said to have multiplicity ``ω``. Multiplicity
+polymorphic functions may have variable multiplicity, *e.g.*
+
+::
+
+  map :: (a:p -> b) -> [a]:p -> [b]
+
+without polymorphism we would need two implementations of `map`. With
+the exact same code: one for ``p=1`` and one for ``p=ω``. Function
+composition is even worse: it takes two multiplicity parameters, hence,
+would require four identical implementations:
+
+::
+
+  (.) :: (b:p -> c) -> (a:q -> b) -> a:(p ':* q) -> c
+
+Syntax
+~~~~~~
+
+The new primary constructs are: multiplicities and the multiplicity
+indexed arrow.
+
+- Multiplicities are a datatype:
+
+  ::
+
+    data Multiplicity
+      = One
+      | Omega
+      | Multiplicity :* Multiplicity
+      | Multiplicity :+ Multiplicity
+
+  In the following, for conciseness, we write ``1`` for ``One`` and
+  ``U`` (ASCII) or ``ω`` (Unicode) for ``Omega``. Note: unification
+  of multiplicities will be performed up to the semiring laws.
+- The multiplicity annotated arrow, for polymorphism, is written ``a:p
+  -> b`` (where ``a`` and ``b`` are types and ``p`` is a
+  multiplicity). To avoid introducing a new notion of "mixfix"
+  operators, we introduce a familly of (infix) type constructors:
+  ``(:p ->)`` for each multiplicity ``p``.
```

Comment:
Wouldn't be more obvious to omit the space between the multiplicity and the arrow? `:p->` instead of `:p ->`?

Reply:
Addressed.